riscv.c

```diff
@@ -770,15 +815,469 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+        return false;
+    }
+    rv->io.mem_write_w(rv, addr, data);
+
+    rv->PC += rv->inst_len;
+    return true;
+}
+
+// CL-type
+static bool op_clw(struct riscv_t *rv, uint16_t inst)
+{
+    uint16_t temp = 0;
+    temp |= (inst & 0b0000000001000000) >> 4;
+    temp |= (inst & FC_IMM_12_10) >> 7;
+    temp |= (inst & 0b0000000000100000) << 1;
+
+    const uint16_t imm = temp;
+    const uint16_t rd = c_dec_rdc(inst) | 0x08;
+    const uint16_t rs1 = c_dec_rs1c(inst) | 0x08;
+    const uint32_t addr = rv->X[rs1] + imm;
+
+    if (addr & 3) {
+        rv_except_load_misaligned(rv, addr);
+        return false;
+    }
+    rv->X[rd] = rv->io.mem_read_w(rv, addr);
+
+    rv->PC += rv->inst_len;
+    return true;
+}
+
+// CS-type
+static bool op_csw(struct riscv_t *rv, uint16_t inst)
+{
+    uint32_t temp = 0;
+    //                ....xxxx....xxxx
+    temp |= (inst & 0b0000000001000000) >> 4;
+    temp |= (inst & FC_IMM_12_10) >> 7;
+    temp |= (inst & 0b0000000000100000) << 1;
+
+    const uint32_t imm = temp;
+    const uint32_t rs1 = c_dec_rs1c(inst) | 0x08;
+    const uint32_t rs2 = c_dec_rs2c(inst) | 0x08;
+    const uint32_t addr = rv->X[rs1] + imm;
+    const uint32_t data = rv->X[rs2];
+
+    if (addr & 3) {
+        rv_except_store_misaligned(rv, addr);
+        return false;
+    }
+    rv->io.mem_write_w(rv, addr, data);
+
+    rv->PC += rv->inst_len;
+    return true;
+}
+
+// CJ-type
+static bool op_cj(struct riscv_t *rv, uint16_t inst)
+{
+    const int32_t imm = (c_dec_cjtype_imm(inst));
+    rv->PC += imm;
+    if (rv->PC & 0x1) {
+        rv_except_inst_misaligned(rv, rv->PC);
+        return false;
+    }
+    // can branch
+    return false;
+}
+
+static bool op_cjal(struct riscv_t *rv, uint16_t inst)
+{
+    const int32_t imm = sign_extend_h(c_dec_cjtype_imm(inst));
+    rv->X[1] = rv->PC + 2;
+    rv->PC += imm;
+    if (rv->PC & 0x1) {
+        rv_except_inst_misaligned(rv, rv->PC);
+        return false;
+    }
+    // can branch
+    return false;
+}
+
+// CR-type
+static bool op_ccr(struct riscv_t *rv, uint16_t inst)
+{
+    const uint32_t rs1 = c_dec_rs1(inst);
+    const uint32_t rs2 = c_dec_rs2(inst);
+    const uint32_t rd = rs1;
+
+    switch ((inst & 0x1000) >> 12) {
+    case 0:
+        if (rs2) {
+            // C.MV
+            rv->X[rd] = rv->X[rs2];
+            rv->PC += rv->inst_len;
+            if (rd == rv_reg_zero)
+                rv->X[rv_reg_zero] = 0;
+        } else {
+            // C.JR
+            rv->PC = rv->X[rs1];
+            return false;
+        }
+        break;
+    case 1:
+        if (rs1 == 0 && rs2 == 0)  // C.EBREAK
+            rv->io.on_ebreak(rv);
+        else if (rs1 && rs2)  // C.ADD
+        {
+            rv->X[rd] = rv->X[rs1] + rv->X[rs2];
+            rv->PC += rv->inst_len;
+            if (rd == rv_reg_zero)
+                rv->X[rv_reg_zero] = 0;
+        } else if (rs1 && rs2 == 0)  // rs1!=zero rs2== zero
+        {
+            // C.JALR
```

Comment:
The comment looks weird. Instead, you can quote from the specification.

Reply:
Should be `} else if (rs1 && rs2 == 0)  { // rs1 != zero AND rs2 == zero`

`{` appears before the comment.